🗒 동기화 메모장 [Step 1] Ryan

ryan-son opened the pull request:
@AppleCEO 

도미닉, 안녕하세요! Step 1 기능 구현을 위한 학습을 하다 PR이 늦었네요. 이제 쉬실 시간인데 리뷰 요청드려 죄송합니다.
오래 걸린만큼 최대한 보시기 편하게 구조를 설계하고 코딩 컨벤션에 맞추어 코드를 작성하는 것에 초점을 맞추었습니다.

## 작동 화면
- iPhone 11 
![ezgif com-video-to-gif (7)](https://user-images.githubusercontent.com/69730931/120777232-0832a080-c560-11eb-9e30-51ccdeac02ab.gif)

![ezgif com-resize (3)](https://user-images.githubusercontent.com/69730931/120778135-f9002280-c560-11eb-94d8-cda8e6a90ee9.gif)

- iPad Pro (12.9-inch) (5th generation)
![image](https://user-images.githubusercontent.com/69730931/120778470-52685180-c561-11eb-8ea3-b3134c21b8c0.png)
![image](https://user-images.githubusercontent.com/69730931/120779242-d91d2e80-c561-11eb-8b56-07457e5a92ce.png)

## 구조 (참고)
폴더 구조는 아래 그림과 같이 구성되어 있습니다.
![image](https://user-images.githubusercontent.com/69730931/120765203-efbc8900-c553-11eb-878b-efd696545bea.png)
각 폴더에는 아래와 같이 파일이 분류가 되어 있습니다.
- `Application`: AppDelegate, SceneDelegate
- `Models`: 모델 타입 및 에러 타입
- `Views` LaunchScreen 및 Cell
- `Controllers`: 뷰 컨트롤러들
- `Assets`: 사용된 리소스. 샘플 JSON 파일
- `Helpers`: 로직 구성에 사용된 프로토콜, 익스텐션, 프레임워크
- `CoreData`: xcdatamodeld 파일

코드가 긴 파일의 경우 구조 파악에 용이하도록 마크주석을 달아두었습니다.
![image](https://user-images.githubusercontent.com/69730931/120775595-4f1f9680-c55e-11eb-870e-1a15ac7b2083.png)

## 요구사항 구현 (참고용)
### 스플릿 뷰
- Regular size class에서 리스트 뷰와 메모 화면을 보여주기 위해`UISplitViewController`의 `init(style:)`을 통해 `doubleColumn` 스타일을 지정해주었습니다.
- `setViewController(_:for:)` 메서드를 통해 리스트뷰를 primary(사이드바), 상세 메모 화면을 secondary로 지정하였습니다.
- `SceneDelegate`의 `scene(_:willConnectTo:options:)` 메서드를 통해 `rootViewController`로 지정하였습니다.
- `NoteShowable` 프로토콜 채택으로 인해 구현한 `showNote(with:)` 메서드를 통해 리스트 뷰의 셀이 선택될 때 리스트 뷰로부터 전달받은 메모 내용을 자신의 하위 뷰컨트롤러로 등록된 상세 메모 화면에 전달합니다.
- 리스트 뷰 셀 선택에 따라 상세 화면을 보여줄 때 `showDetailViewController(_:sender:)` 메서드를 통해 기기의 size class에 따라 화면을 전환합니다.
- `splitViewController(_:topColumnForCollapsingToProposedTopColumn)` 메서드를 통해 size class가 compact인 기기가 앱 실행 시 처음으로 보여줄 화면을 리스트뷰로 설정하였습니다.

### 목록 형식 뷰(리스트뷰)
- `UICollectionView`를 사용하였습니다. 
  - `UICollectionViewCompositionalLayout`을 통해 간편하게 리스트 뷰를 구현할 수 있게 되었으며, 리스트뷰의 형식도 자유롭게 변경할 수 있기 때문에 컬렉션뷰로 구현하였습니다.
- 데이터를 보여주고 화면에 적용하는 방법으로 `UICollectionViewDiffableDataSource`와 `NSDiffableDataSourceSnapshot`을 사용하였습니다.
  - `DiffableDataSource`를 이용하면 각 셀을 구성할 때 해시값을 통해 특정 아이템을 탐색하므로 종전의 indexPath를 이용한 방식보다 탐색 속도가 빠르고, 데이터에 변화가 생겼을 때 편리하게 애니메이션으로 변화를 보여줄 수 있다는 장점 때문에 이 방식을 택했습니다.
- 날짜 형식은  사용자의 지역 설정 기반(`Locale.current`) 보다 `Locale.autoupdatingCurrent` 프로퍼티를 사용해 개인 설정에 기반하여 적용할 수 있도록 적용하였습니다.
- 새 메모를 생성할 수 있도록 추가 버튼을 구현하였습니다.

### 상세 메모화면
- 화면에는 `TextView`와 `NavigationBar`가 위치합니다.
- 기기의 size class가 compact인 경우 리스트 뷰로부터 show 방식으로 화면이 전환됩니다.
- `push`, `pop`을 통해 화면을 쌓지 않고, 텍스트뷰의 내용을 입력받은 데이터로 변경하는 형식으로 동작합니다. 

## 조언을 얻고 싶은 부분 / 해결하지 못한 점
### 1. 텍스트 뷰를 선택하고 다른 셀을 선택할 때 텍스트 뷰의 커서(편집 활성화 상태)가 없어지지 않습니다.
이 원인 때문인지 아이폰에서는 키보드 활성화 상태에서 백 버튼으로 돌아간 뒤 다른 셀을 눌러 상세 메모 화면으로 진입하면 진입하자마자 키보드가 활성화됩니다. 아이패드에서는 textView의 resignFirstResponder 메서드를 통해 키보드가 활성화되지 않도록 하였지만 실제 기기가 없어 실제로 키보드가 사라지는지는 테스트해보지 못했습니다. 시뮬레이터 상에서는 아래와 같은 편집용 키보드가 활성화되지는 않음을 확인하였습니다.
![image](https://user-images.githubusercontent.com/69730931/120780506-2a79ed80-c563-11eb-8b3c-4e9b0c8378c8.png)

이를 해결하기 위한 방법이나 참고할 수 있는 자료가 있을까요?

#### 예시 화면
- 앱을 실행하고 처음 셀을 선택했을 때 (커서 없음)
![image](https://user-images.githubusercontent.com/69730931/120773226-1a124480-c55c-11eb-8b2d-50de97f3af92.png)

- 텍스트뷰 첫 선택 (커서 생김)
![image](https://user-images.githubusercontent.com/69730931/120774018-d409b080-c55c-11eb-9738-efee9e138a68.png)

- 다른 셀을 선택하여 이동했을 때 (마지막 글자에 커서 유지)
![image](https://user-images.githubusercontent.com/69730931/120775926-a32a7b00-c55e-11eb-90e1-d670dedc44de.png)

### 아이패드에서 키보드 활성화 시 오토레이 아웃 관련 에러가 발생하지만 해결하지 못했습니다.
해결하기 위한 방법을 어떻게 알아보아야 할지 알려주시면 감사하겠습니다!
- 앱 실행시 콘솔 (빈 화면)
![image](https://user-images.githubusercontent.com/69730931/120779579-27cac880-c562-11eb-87fd-06b0ab40df7f.png)

- 키보드 활성화 시 콘솔
![image](https://user-images.githubusercontent.com/69730931/120779720-5052c280-c562-11eb-80b1-bf5d6b8098db.png)
  - 키보드 활성화 화면
  - ![image](https://user-images.githubusercontent.com/69730931/120779800-6791b000-c562-11eb-8dbd-e42469432d4a.png)

에러 코드를 wtfautolayout.com에 입력한 결과. 코드에도 나타나있듯이 바 버튼의 오토레이아웃을 조정하지는 않았습니다.
![image](https://user-images.githubusercontent.com/69730931/120780783-64e38a80-c563-11eb-84a6-9241265f9c19.png)

주말 마무리 잘 하시기 바랍니다~ 감사합니다! 😄

ryan-son committed:
Update issue templates

ryan-son committed:
Merge pull request #1 from ryan-son/ryan-son-patch-1

Update issue templates

ryan-son committed:
chore: Add `SwiftLint` for dependency

- Minor setting included

ryan-son committed:
chore: Set project folder structure

- [Reference](https://stackoverflow.com/questions/39945727/best-practice-for-an-xcode-project-groups-structure)

ryan-son committed:
chore: Switch method for initiating application

- Switched from storyboard to SceneDelegate to initiate application

ryan-son committed:
chore: Add pods folder into gitignore

ryan-son committed:
chore: Exclude files not WIP from linting subjects

ryan-son committed:
chore: Add sample asset for testing

ryan-son committed:
feat: Build a model for `Note`

ryan-son committed:
test: Add test method for verifying `Note` model

ryan-son committed:
feat: Implement listView with collectionView

- Implemented list view with UICollectionViewCompositionalLayout
- Implemented cell for collection view
- Made `UIError` and `DataError` to express error
- Test method for verifying the extended initializer of `JSONDecoder` included
- Test method for verifying `loadNotes(from:)`

ryan-son committed:
fix: Fix bug for `identifiers are not unique`

- Added UUID to `Note` to have unique hash value
- Changed the initializer of JSONDecoder extension due to unrequired prop set.

Resolves: #3

ryan-son committed:
fix: adjust cell UI constraints and add accessory

- Add disclosure indicator for accessory

ryan-son committed:
feat: Add and apply date formatting feature

ryan-son committed:
feat: Implement deselectItem method for cells

ryan-son committed:
refactor: Refactor with `loadNotes(from:)`

- Separated responsibilities for getting asset and decoding tasks
- Tasks above are now responsible to `decode(to:from:)` in JSONDecoder
- `loadNotes(from:)` can concentrate its own job

ryan-son committed:
Merge pull request #6 from ryan-son/feature/step1/listView

Feature/step1/list view

ryan-son committed:
feat: Implement split view

- Initiated with `SceneDelegate`
- Added `NoteSplitViewController` and `NoteDetailViewController`
- `NoteDetailViewController` will show detailed note for the selected cell

ryan-son committed:
style: Remove unnecessary comments and spaces

ryan-son committed:
feat: Reinforce error description with associated values

ryan-son committed:
feat: Add `NoteDetailViewController` features

- shows selected note content

ryan-son committed:
refactor: Remove raw strings in detail view controller

ryan-son committed:
refactor: refactor with methods and some constants

ryan-son committed:
feat: Add features for navigation bar and adding note

ryan-son committed:
feat: Change logic for adding a new note

ryan-son committed:
chore: Set access level to methods

- Set access level to methods missing from previous commit

ryan-son committed:
feat: Change logic for passing note data to secondary

`noteSplitViewControllerDelegate` in `NoteListViewController` now performs
data passing to secondary view and showing.
`collectionView(_:didSelectItemAt:)` method only sends splitViewDelegate to
show note.

ryan-son committed:
Merge pull request #7 from ryan-son/feature/step1/splitView

Feature/step1/split view

AppleCEO commented on CloudNotes/CloudNotes/Controllers/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,129 @@
+
+import UIKit
+
+final class NoteDetailViewController: UIViewController {
```
final을 붙여준 이유는 무엇인가요?

ryan-son replied:
해당 클래스가 상속을 통해 재사용되는 일이 없을 것이라 판단하여 `final` 키워드를 사용하였습니다. `final`로 선언된 메서드들은 직접 호출하는 반면 그렇지 않은 요소들은 vtable을 통해 간접 호출되어 직접 호출되는 경우보다 느리게 작동한다는 내용을 참고하여 적용하였습니다.
- [Writing High-Performance Swift Code -Apple Swift Github](https://github.com/apple/swift/blob/main/docs/OptimizationTips.rst#advice-use-final-when-you-know-the-declaration-does-not-need-to-be-overridden)
- [Understanding Swift Performance -WWDC16](https://developer.apple.com/videos/play/wwdc2016/416/?time=1374)

AppleCEO commented on CloudNotes/CloudNotes/Controllers/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,129 @@
+    }()
+    
+    // MARK: - Namespaces
+    private enum ViewContents {
+        /// shows when the screen first loaded with regular size class.
+        static let welcomeGreeting = "환영합니다!"
+        static let newLineString = "\n"
+        static let emptyString = ""
+    }
```
열거형을 활용한 부분 좋습니다! 👍

AppleCEO commented on CloudNotes/CloudNotes/Controllers/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,129 @@
+        configureNavigationBar()
+        configureTextView()
+        setNoteBackgroundColor(to: .systemBackground)
+        updateUI()
```
추상화 단계가 같은 코드들이 모여있어서 보기 좋습니다~ 🥇

AppleCEO commented on CloudNotes/CloudNotes/Controllers/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,129 @@
+    }
+    
+    func updateUI() {
+        guard let note = note else {
+            setGreetingText(to: noteTextView)
+            return
+        }
```
기본형을 사용한 부분 보기 좋습니다~ 여기 if let을 사용하면 어땠을까요?

ryan-son replied:
도미닉이 제안해주신 내용이 더 가독성이 좋다고 생각합니다. 또, `guard`문을 통한 빠른 종료는 뷰를 최초에 표현할 때만 적용되므로 아래와 같이 작성하는 것이 더 나은 선택일 수 있겠네요!

```swift
func updateUI() {
    if let note = note {
        setText(to: noteTextView, with: note)
    } else {
        setGreetingText(to: noteTextView)
    }
}
```

AppleCEO replied:
저는 이런 방식을 선호합니다.
```swift
func updateUI() {
    if let note = note {
        setText(to: noteTextView, with: note)
        return
    }

    setGreetingText(to: noteTextView)
}
```

AppleCEO commented on CloudNotes/CloudNotes/Controllers/NoteListViewController.swift:
```diff
@@ -0,0 +1,191 @@
+}
+
+// MARK: - Cell Nib Registration
+extension NoteListViewController {
+    private func registerCellNib() {
+        let noteCellNib = UINib(nibName: NoteCollectionViewListCell.reuseIdentifier, bundle: .main)
+        noteListCollectionView?.register(
+            noteCellNib,
+            forCellWithReuseIdentifier: NoteCollectionViewListCell.reuseIdentifier
+        )
+    }
+}
+
+// MARK: - Create Layout for Collection View
+extension NoteListViewController {
+    private func createLayout() -> UICollectionViewLayout {
+        let configuration = UICollectionLayoutListConfiguration(appearance: .sidebarPlain)
+        return UICollectionViewCompositionalLayout.list(using: configuration)
+    }
+}
```
extension 을 각각 나눈 이유가 있을까요? 🔢

ryan-son replied:
- scope을 기능 단위별로 분리하길 희망하여 복수의 extension으로 나누었습니다. 이러한 방법을 사용하면 extension을 반복해서 작성하여야 하지만, 추후에 메서드의 단위를 분리하거나 요구사항이 변경되어 각 기능 단위별로 메서드를 추가할 경우가 생기더라도 단순히 mark 주석으로 단위를 구분하는 것보다 기능 단위별로 관련성이 높은 메서드별로 구분할 수 있고, 향후 코드를 유지보수할 때도 기능 단위별로 scope가 나뉘어 있으니 다른 프로그래머가 코드를 보더라도 복수의 extension을 선언한 의미를 파악할 수 있을 것이라 생각했습니다.
- 프로토콜 채택을 통해 구현하는 메서드가 있는 경우 [extension으로 구분하여 작성하는 경우](https://github.com/raywenderlich/swift-style-guide#code-organization)가 있는 것으로 파악했는데, 코멘트 해주신 부분처럼 단순히 기능 관련성이 높은 메서드별로 extension을 구분하는 방법은 불필요한 extension 구문을 반복해서 작성해야 하기 때문에 좋지 않은 방법일 수 있다고 생각합니다. 도미닉의 의견은 어떠신가요?

AppleCEO replied:
메소드마다 하나의 extension은 좋지 않다는 의견입니다. 
관련된 메소드끼리 묶던지 아니면 아무런 프로토콜 채택을 하지 않는 메소드들을 하나의 extension에 모으는 것도 괜찮지 않을까라고 생각합니다.

ryan-son replied:
의견 주신 부분을 반영하여 프로토콜과 관련없는 extension들은 본체의 클래스 또는 구조체로 이동하게끔 하였습니다. 불필요하게 구분하였을 때보다 보기가 더 편해진 것 같네요!

AppleCEO commented on CloudNotes/CloudNotes/Controllers/NoteListViewController.swift:
```diff
@@ -0,0 +1,191 @@
+extension NoteListViewController {
+    private func configureNavigationBar() {
+        navigationItem.title = NavigationBarItems.title
+        
```
여기 공백은 어떤 의미일까요?

ryan-son replied:
불필요한 곳에 공백을 삽입했네요. 수정하도록 하겠습니다!

AppleCEO commented on CloudNotes/CloudNotes/Controllers/NoteListViewController.swift:
```diff
@@ -0,0 +1,191 @@
+    
+    @objc private func addTapped() {
+        guard let dataSource = dataSource else {
+            os_log(.fault, log: .data, OSLog.objectCFormatSpecifier, DataError.dataSourceNotSet.localizedDescription)
```
위에서는 여러 줄로 나열해서 파라미터를 전달했던 것 같은데 여기는 한 줄이군요!
통일하는 것이 좋겠습니다.
저는 보통 이 줄처럼 한 줄로 작성하는 것을 선호합니다.

ryan-son replied:
- 저는 한 라인이 120 자 즈음을 초과할 때 (정확히는 현재 배율에서 화면을 벗어나 강제 line-wrapping이 발생할 때)를 기준으로 줄바꿈을 하고 있습니다.
- 도미닉의 의견처럼 동일한 메서드를 호출할 때 동일한 기준으로 줄바꿈하는 것이 가독성 측면에서 더 유리하다고 판단하여 `os_log` 메서드의 경우에는 줄바꿈 정책을 한 줄로 동일하게 변경하였는데, 메서드를 정의하거나 호출하는 부분에서의 줄바꿈 정책도 한 줄로 변경해야할지에 대한 고민을 하고 있습니다.
- 아래 캡처와 같이 유사한 메서드들의 호출이지만 사용자의 화면 배율 또는 해상도에 따라 강제 line-wrapping이 발생하여 가독성이 떨어질 수 있는 경우 두 번째 캡처와 같이 일관된 줄바꿈 방식을 적용하여 표현하는 것이 좋지 않을까 생각하는데, 도미닉의 의견이 궁금합니다.
![image](https://user-images.githubusercontent.com/69730931/120920936-4369d680-c6fc-11eb-8b34-4e016fc17e99.png)
![image](https://user-images.githubusercontent.com/69730931/120920986-73b17500-c6fc-11eb-8695-823ede13a7f2.png)

AppleCEO replied:
저는 위에 캡처가 그래도 낫다는 의견입니다. 연속성있게 읽기 좋을 것 같아요. 
바로 다음에 constant가 오는데 줄이 길어져서 아래로 내려갔구나 예상할 수 있는 것 같습니다. 
반면에 아래는 이게 초기화가 끝난 것이지 함수가 실행되는 것인지 조금 가독성이 떨어지는 것 같아요.

ryan-son replied:
함수 정의 및 호출과 관련된 코드 전체의 줄바꿈 스타일을 변경하였습니다!

AppleCEO commented on CloudNotes/CloudNotes/Controllers/NoteSplitViewController.swift:
```diff
@@ -0,0 +1,55 @@
+}
+
+extension NoteSplitViewController: UISplitViewControllerDelegate {
+    func splitViewController(
+        _ svc: UISplitViewController,
+        topColumnForCollapsingToProposedTopColumn proposedTopColumn: UISplitViewController.Column
```
이 부분도 여러 줄이어서 저는 오히려 보기 힘든 것 같습니다. 어떻게 생각하시나요?

ryan-son replied:
프로토콜의 지원에 따라 구현할 수 있는 메서드는 메서드의 종류만 정확히 파악할 수 있으면 문제가 되지 않는다고 생각하기에 줄바꿈을 통해 매개변수 별로 구분해주지 않아도 괜찮다고 생각합니다. 도미닉 덕분에 일관적으로 기준을 적용하기 보다 상황에 따라 줄바꿈 기준을 달리할 수 있다는 생각을 하게 되었습니다. 감사합니다! 😄

AppleCEO commented on CloudNotes/Podfile:
```diff
@@ -0,0 +1,20 @@
+target 'CloudNotes' do
+  # Comment the next line if you don't want to use dynamic frameworks
+  use_frameworks!
+  pod 'SwiftLint'
```
스위프트 린트를 사용해보셨군요! 💯

ryan-son replied:
도미닉의 피드백 덕분에 스위프트 린트가 제시하는 기준에 매몰되지 않고 상황에 따라 적절한 기준으로 코드의 형식을 정할 수 있게 되었습니다. 앞으로도 말씀해주신 내용 잘 기억해서 도구에 매몰되지 않도록 하겠습니다~!

ryan-son committed:
chore: Remove unnecessary initializer from `Note`

- Use memberwise initializer instead

ryan-son committed:
chore: Correct typo

AppleCEO reviewed (commented):
1. 텍스트 뷰를 선택하고 다른 셀을 선택할 때 텍스트 뷰의 커서(편집 활성화 상태)가 없어지지 않습니다.
제 생각에는 같은 뷰컨트롤러에서 내용만 바뀌고 있는 상황 같습니다. 편집 비활성화 상태로 바꾸는 코드를 viewwillappear에 넣어주면 어떨까요?
2. 아이패드에서 키보드 활성화 시 오토레이 아웃 관련 에러가 발생하지만 해결하지 못했습니다.
오토레이아웃 문제로 보이네요! 오토레이아웃은 민감해서 조금만 맞지 않으면 오류가 뜨곤 합니다. 어떤 부분이 문제일지 찾아보셔야 할 것 같습니다. 바 버튼과 관련된 문제로 보이는데 어떤 부분을 찾아보셨는지 공유해주실 수 있을까요?

ryan-son commented:
@AppleCEO 

도미닉, 주말은 잘 보내셨나요? 이르게 리뷰해주셨는데 제가 코멘트를 늦게 드려 죄송한 마음입니다. 🥲

>1. 텍스트 뷰를 선택하고 다른 셀을 선택할 때 텍스트 뷰의 커서(편집 활성화 상태)가 없어지지 않습니다.
> 제 생각에는 같은 뷰컨트롤러에서 내용만 바뀌고 있는 상황 같습니다. 편집 비활성화 상태로 바꾸는 코드를 viewwillappear에 넣어주면 어떨까요?
> 아이패드에서 키보드 활성화 시 오토레이 아웃 관련 에러가 발생하지만 해결하지 못했습니다.

- 아래 코드를 `viewWillAppear(_:)`과 `viewWillDisappear(_:)`에 삽입하여 문제를 해결하였습니다. 도미닉이 의도하신 내용이 아래 내용이 맞나요?
```swift
override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(true)
    noteTextView.isEditable = false
}

override func viewDidAppear(_ animated: Bool) {
    super.viewDidAppear(true)
    noteTextView.isEditable = true
}
```

> 2. 오토레이아웃 문제로 보이네요! 오토레이아웃은 민감해서 조금만 맞지 않으면 오류가 뜨곤 합니다. 어떤 부분이 문제일지 찾아보셔야 할 것 같습니다. 바 버튼과 관련된 문제로 보이는데 어떤 부분을 찾아보셨는지 공유해주실 수 있을까요?

커뮤니티에는 최근까지 해결방법을 찾지 못한 것으로 확인하여 ([글1](https://developer.apple.com/forums/thread/667441), [글2](https://stackoverflow.com/questions/64895600/uibuttonbarbutton-layout-warning-on-entering-uitextview-on-ipad-not-on-iphone), [글3](https://www.hackingwithswift.com/forums/ios/uibuttonbarbutton-layout-warning-on-entering-uitextview-on-ipad-not-on-iphone/4882)) 앞으로 어떻게 해결 방법을 알아보아야 할지 질문드렸습니다. 제가 아직 파악하지 못한 내용도 반드시 있을 것이라 생각하는데, 만약 있으시면 도미닉이 파악하고 계신 내용을 공유해주실 수 있으실까요? 없으시다면 커뮤니티 의견처럼 아직까지 명확한 해결 방법이 없고 애플리케이션 작동에 큰 문제가 없으니 그대로 두는 것 외에는 대응방안이 없을지 궁금합니다.

## 추가 내용
### `splitView`가 하위 뷰인 `detailView`에게 노트를 보여달라는 메시지를 보내는 과정을 추상화하였습니다.
기존 구조는 아래와 같이 상위 뷰인 `splitView`가 하위 뷰인 `detailView`에게 과도하게 세부적인 요구를 하고 있다고 판단하여 인터페이스를 더 추상화하였습니다. 이를 통해 노트를 보여줄 책임이 있는 `detailView`가 기존보다 더 자유롭게 상위뷰의 메시지에 대한 책임을 완수할 수 있도록 구성하였습니다. 기존보다 더 나은 방식으로 구성한 것이 맞을까요? 도미닉의 의견은 어떠신가요?
```swift
// 기존 구조
extension NoteSplitViewController: NoteShowable {
    func showNote(with note: Note) {
        guard let noteDetailViewController = self.viewController(for: .secondary) as? NoteDetailViewController else {
            os_log(.error, log: .ui, OSLog.objectCFormatSpecifier, UIError.downcastingFailed("Secondary view controller", #function).localizedDescription)
            return
        }
        noteDetailViewController.setContent(with: note) // listView로 부터 전달 받은 데이터를 줄테니 세팅해라
        noteDetailViewController.updateTextView() // 가지고 있는 텍스트 뷰의 텍스트를 전달한 컨텐츠로 업데이트 해라
        noteDetailViewController.noteTextView.resignFirstResponder() // 활성화된 키보드를 제거해라
        showDetailViewController(noteDetailViewController, sender: nil)
    }
}

// 개선한 구조
extension NoteSplitViewController: NoteShowable {
    func showNote(with note: Note) {
        guard let noteDetailViewController = self.viewController(for: .secondary) as? NoteDetailViewController else {
            os_log(.error, log: .ui, OSLog.objectCFormatSpecifier, UIError.downcastingFailed("Secondary view controller", #function).localizedDescription)
            return
        }
        noteDetailViewController.showContent(with: note) // listView에서 전달 받은 컨텐츠를 보낼테니 알아서 보여주어라
        showDetailViewController(noteDetailViewController, sender: nil)
    }
}

// 개선한 구조와 함께 더 자유롭게 구현 가능해진 detailView의 showContent(with:) 메서드
func showContent(with inputNote: Note) { // 상위 뷰로부터 전달받은 컨텐츠로 화면을 보여달라는 메시지를 수신했다.
    note = inputNote // 전달받은 컨텐츠를 프라이빗 프로퍼티에 할당하자.
    updateTextView() // 텍스트 뷰의 내용을 전달 받은 컨텐츠로 최신화하자.
    moveTop(of: noteTextView) // 컨텐츠를 보여달라고 할 때마다 텍스트 뷰의 가장 최상단에 위치시켜주자.
    removeActivatedKeyboard() // 키보드가 활성화되어 있다면 없애주자.
}
```

### collectionView의 셀 선택에 따라 textView로 진입할 때 항상 작성한 첫 라인이 보이도록 뷰를 배치하였습니다.
아래 메서드를 통해 기능을 구현하였는데, `viewWillAppear(_:)` 메서드와 같은 View Life Cycle에 해당 메서드를 추가하지 않은 이유는 regular size class에서는 노트의 내용을 보여주는 detail View의 view life cycle의 메서드가 매 번 실행되지 않기 때문입니다. 그 대신, 상위 뷰인 `splitView`가 새로운 노트를 보여달라는 메시지를 송신하여 `detailView`가 메시지를 수신하면 뷰가 항상 텍스트 뷰의 최상단에 위치하도록 조정할 수 있도록 `detailView`의 `showContent(with:)` 메서드에 해당 기능을 삽입하였습니다.
```swift
private func moveTop(of textView: UITextView) {
    textView.setContentOffset(CGPoint(x: 0, y: -view.safeAreaInsets.top), animated: false)
}
```
- `CGPoint`의 y방향을 0으로 지정하였을 때는 내비게이션 바에 텍스트가 일정 부분 가리는 현상이 있어 [이 글](https://stackoverflow.com/questions/34670556/why-my-uitextview-text-is-not-show-from-starting-and-hide-under-navigation-bar)을 참고하여 `-view.safeAreaInsets.top` 만큼 오프셋을 설정하였습니다.
- [이 글](https://www.hackingwithswift.com/example-code/uikit/how-to-stop-your-view-going-under-the-navigation-bar-using-edgesforextendedlayout)에 따르면 이런 이슈는 iOS 7.0부터 애플이 'context (정확히 어떤 내용인지 이해하지 못했습니다)'를 제공하기 위해 내비게이션 바와 탭 바 뒤로 일정 부분의 UI가 숨겨지게끔 제작하였다는데 저는 정확히 파악하지는 못했습니다. 관련해서 도미닉이 알고 계신 내용이 있으시다면 공유해주시면 감사하겠습니다. 링크된 내용의 해결 방법`edgesForExtendedLayout = []`을 이용하지 않은 이유는 뷰의 시작 위치가 간헐적으로 잘못되는 경우가 있었기 때문입니다 (아래 그림).
![image](https://user-images.githubusercontent.com/69730931/120923962-51275800-c70c-11eb-931a-aec388e0fc69.png)

### Step 1을 진행하고 있는 만큼 초기 단계 구조를 잘 잡는 것이 중요하다고 생각합니다. 현재 구조에서 개선하여야 할, 또는 개선하면 좋을 부분이나 왜 이렇게 만들었는지 의문스러운 부분이 있을까요?

ryan-son committed:
fix: Fix keyboard activation and textView UI bugs

Fixed items:
1. Keyboard activates when selects textView once and moves to other detail views
2. TextView does not start with first line of the text

ryan-son committed:
refactor: Adjust showing content responsibilities

- Now `NoteDetailViewController` has all responsibilities for showing content
- It is to guarantee the autonomy of `NoteDetailViewController` for
  its responsibilies of showing note content

ryan-son committed:
chore: Correct typo and add parameter names to error

ryan-son committed:
chore: Remove unnecessary nib life cycle method

- `awakeFromNib()` is only called when the cell is loaded from xib or storyboard

AppleCEO commented:
아래 코드를 viewWillAppear(_:)과 viewWillDisappear(_:)에 삽입하여 문제를 해결하였습니다. 도미닉이 의도하신 내용이 아래 내용이 맞나요?
- 네 맞습니다! 💯 

제가 아직 파악하지 못한 내용도 반드시 있을 것이라 생각하는데, 만약 있으시면 도미닉이 파악하고 계신 내용을 공유해주실 수 있으실까요?
- 저도 처음보는 에러이고 보내주신 링크들을 보니 무시하고 넘어갔다는 내용이 대부분이군요. 
- 혹시 질문 답변 채널에 질문 한번 남겨보면 어떨까요?

splitView가 하위 뷰인 detailView에게 노트를 보여달라는 메시지를 보내는 과정을 추상화하였습니다.
- 이 경우 note라는 프로퍼티를 뷰컨트롤러가 가지고 있고 이 프로퍼티 내용대로 뷰컨트롤러가 보여지면서 띄우는 형식이기 때문에 splitView가 detailView에게 데이터만 전달해주고 detailView는 띄어지면서(viewWillAppear) 데이터를 보여주는 형식이 나을 것 같습니다.

collectionView의 셀 선택에 따라 textView로 진입할 때 항상 작성한 첫 라인이 보이도록 뷰를 배치하였습니다.
- 보내주신 링크 중에 edgesForExtendedLayout 는 [링크](https://stackoverflow.com/questions/18798792/explaining-difference-between-automaticallyadjustsscrollviewinsets-extendedlayo)를 보시면 레이아웃의 끝을 어디까지로 볼거인지의 대한 내용으로 생각됩니다. textView로 진입할 때 처음이 보이지 않는 문제와는 상관없는 것 같습니다. 
- 네비게이션 뷰에서는 뷰컨트롤러 객체가 한번 생성되고 뒤로갔을 때 NoteDetailViewController 객체가 사라지지는 않습니다. 그래서 디버깅해보시면 처음 NoteDetailViewController 를 열었을 때는 ViewDidLoad가 실행되지만 뒤로 가기했다가 다시 열면 NoteDetailViewController의 ViewDidLoad가 실행되지는 않습니다. 
- 내용만 바뀌면서 내려간 스크롤의 상태도 저장이 되어있다가 다음 NoteDetailViewController가 열릴 때 그 위치로 가는 문제로 보입니다. 
- 현재 방식대로 해결하신 것도 좋지만 뷰가 다시 열릴 때 note 변수인 내용을 가지고 텍스트뷰의 내용을 바꾸고 위로 전환시키는 것이 어떨까라고 생각합니다. 
- noteDetailViewController 에 note 값에 내용을 할당해주고 viewWillAppear에서 note의 값대로 텍스트뷰를 보여주는 것이 어떨까요? 그리고 그 때 처음으로 이동시키는거죠.

위에 구조 한번 고민해보시고 이해 안되는 부분 있으면 말씀해주세요~

AppleCEO commented:
https://www.boostcourse.org/mo326/lecture/16894?isDesc=false
이 구조 한번 참고해주세요!

ryan-son committed:
style: Change code-dividing policy

ryan-son committed:
style: Merge private methods from extensions to main

ryan-son committed:
refactor: Change data passing structures

- Now selected note data is passed from list view to detail view directly.
- This change is to keep data passing structures simple now and after.

ryan-son commented:
@AppleCEO 

도미닉, 점심시간에 질문드렸는데 신속하게 답변해주셔서 감사드립니다.

논의한 내용과 추가 피드백 주신 부분을 적용하여 코멘트를 남깁니다.

## 데이터 전달 구조 변경
[논의한 구조](https://www.raywenderlich.com/4613809-uisplitviewcontroller-tutorial-getting-started)를 적용하여 데이터 이동이 일어나는 구조를 아래와 같이 변경하였습니다. 구조를 수정한 이유는 논의한 바와 같이 향후 텍스트뷰에서 수정한 내용을 디테일 뷰와 디테일 뷰 데이터에 반영해야할 때 데이터 이동 구조가 복잡해질 수 있기 때문입니다.
- 기존: listView -> splitView -> detailView
- 수정: listView -> detailView

## 줄바꿈 스타일
제안해주신 방식을 채택하여 메서드 정의 및 호출과 관련된 코드에서 전달인자 별 개행 방식을 제거하고 하나의 라인으로 표현하는 방식으로 변경하였습니다.

## 프로토콜을 채택하지 않은 extension의 병합
프로토콜 채택에 따라 구현된 메서드만 extension으로 구분하였고, 그렇지 않은 메서드들은 본체의 클래스 또는 구조체로 이동하게끔 하였습니다.

빠뜨린 부분이 있다면 말씀해주시면 감사하겠습니다. 오늘도 즐거운 하루 보내세요~! 😄

ryan-son committed:
refactor: Change updating text view code style

ryan-son committed:
style: Change line-diving policy - add missed

AppleCEO commented:
훨씬 깔끔해진 느낌이 듭니다! 
요구사항들도 잘 만족하는 것 같습니다. 머지를 원하시나요?
컨벤션을 제가 말씀드린 대로 거의 수정하셨는데 제가 꼭 맞다는 것은 아닙니다 ㅎㅎ 
참고해주시고 더 좋은 방식이 있다고 생각하면 말씀해주세요~

ryan-son commented:
@AppleCEO 

가능하시다면 머지 요청 드립니다! 도미닉과 맞춘 컨벤션으로 더 보기 좋은 코드로 돌아오겠습니다. 정성스러운 리뷰 감사드려요! 😄

AppleCEO commented:
고생하셨습니다!
